app/controllers/benefits/CompanyBenefitController.scala

```diff
@@ -113,20 +114,35 @@ class CompanyBenefitController @Inject()(
         }
       },
       success => {
-        success match {
-          case Some(NoIDontGetThisBenefit) =>
-            journeyCacheService.cache(DecisionChoice, NoIDontGetThisBenefit) map { _ =>
-              Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.stopDate())
-            }
-          case Some(YesIGetThisBenefit) =>
-            journeyCacheService.cache(DecisionChoice, YesIGetThisBenefit) map { _ =>
-              Redirect(
-                controllers.routes.ExternalServiceRedirectController
-                  .auditInvalidateCacheAndRedirectService(TaiConstants.CompanyBenefitsIform)
-                  .url)
+        val benefitType = journeyCacheService.mandatoryJourneyValueAs[String](
+          EndCompanyBenefit_BenefitTypeKey,
+          x => x
+        )
+
+        benefitType.flatMap {
+          case Right(name) =>
+            success match {
+              case Some(NoIDontGetThisBenefit) =>
+                journeyCacheService
+                  .cache(getBenefitDecisionKey(Some(name), DecisionChoice), NoIDontGetThisBenefit) map { _ =>
+                  Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.stopDate())
+                }
+              case Some(YesIGetThisBenefit) =>
+                journeyCacheService.cache(getBenefitDecisionKey(Some(name), DecisionChoice), YesIGetThisBenefit) map {
+                  _ =>
+                    Redirect(
+                      controllers.routes.ExternalServiceRedirectController
+                        .auditInvalidateCacheAndRedirectService(TaiConstants.CompanyBenefitsIform)
+                        .url)
+                }
             }
+
+          case Left(e) => throw new NoSuchElementException(e)
```

Comment:
Don't throw exceptions.
Prefer return types and logging - here you could have return a status like `InteralServerError` but it really depends on your use case.

Reply:
Redirecting instead